amaranth_cli/__init__.py

```diff
@@ -0,0 +1,126 @@
+        metavar="COMPONENT", help="Amaranth component to convert, e.g. `pkg.mod:Cls`",
+        dest="component", type=component)
+    op_generate.add_argument(
+        "-p", "--param", metavar=("NAME", "VALUE"), help="parameter(s) for the component",
+        dest="params", nargs=2, type=str, action="append", default=[])
+    gen_language = op_generate.add_subparsers(
+        metavar="LANGUAGE", help="language to generate code in",
+        dest="language", required=True)
+
+    lang_verilog = gen_language.add_parser(
+        "verilog", help="generate Verilog code")
+    lang_verilog.add_argument(
+        "-v", metavar="VERILOG-FILE", help="Verilog file to write",
+        dest="verilog_file", type=argparse.FileType("w"))
+    lang_verilog.add_argument(
+        "-d", metavar="DEP-FILE", help="Make-style dependency file to write",
+        dest="dep_file", type=argparse.FileType("w"))
+
+    return parser
+
+
+def main(args=None):
+    # Hook the `open()` function to find out which files are being opened by Amaranth code.
+    files_being_opened = set()
+    special_file_opened = False
+    def dep_audit_hook(event, args):
+        nonlocal special_file_opened
+        if files_being_opened is not None and event == "open":
+            filename, mode, flags = args
+            if mode is None or "r" in mode or "+" in mode:
+                if isinstance(filename, bytes):
+                    filename = filename.decode("utf-8")
+                if isinstance(filename, str) and stat.S_ISREG(os.stat(filename).st_mode):
+                    files_being_opened.add(filename)
+                else:
+                    special_file_opened = True
+    sys.addaudithook(dep_audit_hook)
+
+    # Parse arguments and instantiate components
+    args = _build_parser().parse_args(args)
+    if args.operation in ("generate", "gen", "g"):
+        params = dict(args.params)
+        params = {name: cls(params[name])
+                  for name, cls in args.component.__init__.__annotations__.items()}
+        component = args.component(**params)
+
+    # Capture the set of opened files, as well as the loaded Python modules.
+    files_opened, files_being_opened = files_being_opened, None
+    modules_after = list(sys.modules.values())
+
+    # Remove *.pyc files from the set of open files and replace them with their *.py equivalents.
+    dep_files = set()
+    dep_files.update(files_opened)
+    for module in modules_after:
+        if getattr(module, "__spec__", None) is None:
+            continue
+        if module.__spec__.cached in dep_files:
+            dep_files.discard(module.__spec__.cached)
+            dep_files.add(module.__spec__.origin)
+
+    if args.operation in ("generate", "gen", "g"):
+        if args.language == "verilog":
+            # Generate Verilog file.
+            if args.verilog_file:
+                from amaranth.back import verilog
+                args.verilog_file.write(verilog.convert(component))
```

Comment:
I think the `verilog.convert` invocation should include name field. In projects where you generate multiple Verilog modules from Amaranth, chances are the synthesizer is not going to like them all named `top`.

Reply:
The problem is that you'll still get duplicate names for the submodules in that case, so it's not enough.

Reply:
I agree, per IRC, that this is a problem that needs to be solved. For the purposes of testing/finishing the refactor of my code for the time being, would you be willing to expose top-level module renaming from the CLI? It would allow me to write the FuseSoC generation in terms of the Amaranth CLI.

Reply:
Try the updated version.

Reply:
This correctly creates the hiearchy:

```sh
pdm run amaranth generate efbutils.ufm.reader.demo:Demo -p num_leds 2         -p efb_config '{ "dev_density": "7000L", "wb_clk_freq": 12.08 }' -p ufm_config '{ "init_mem": null, "zero_mem": true, "start_page": 0, "num_pages": 2046}' verilog -v - | grep "module "
module top(rx, tx, leds);
module \top.efb (bus__adr, bus__cyc, bus__dat_r, bus__dat_w, bus__irq, bus__stb, bus__we, rst, clk, bus__ack);
module \top.por (clk, por_clk, rst);
module \top.reader (bus__data, bus__read_en, bus__valid, efb__ack, efb__adr, efb__cyc, efb__dat_r, efb__dat_w, efb__irq, efb__stb, efb__we, rst, clk, bus__addr);
module \top.reader.pagemod (clk, seq__data, seq__addr, seq__stb, seq__ack, rand__data, rand__addr, rand__read_en, rand__valid, rst);
module \top.reader.streammod (clk, efb__ack, efb__adr, efb__cyc, efb__dat_r, efb__dat_w, efb__irq, efb__stb, efb__we, stream__data, stream__addr, stream__stb, stream__ack, stream__ready, rst);
module \top.reader.streammod.seqmod (clk, ctl__cmd, ctl__data_len, ctl__done, ctl__op_len, ctl__rd__data, ctl__rd__stb, ctl__req, ctl__wr__data, ctl__xfer_is_wr, efb__ack, efb__adr, efb__cyc, efb__dat_r, efb__dat_w, efb__stb, efb__we, rst);
module \top.uart (i, data, ack, rdy, rst, clk, o);
module \top.uart.rx (rst, clk, divisor, i);
module \top.uart.tx (data, ack, rdy, rst, clk, divisor, o);
module \top.wait (rst, clk, stb);
```

Once I'm able to rename the `top` prefix (probably via an `-n [NAME]` argument to the CLI?), I can finish the refactor, and wait for other ppl to test.

Reply:
Try the updated version.